org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/snippets/PrologSnippetContext.java

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+* 
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.services.snippets;
+
+import java.util.Map;
+
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.services.extensions.ICompletionRequest;
+import org.eclipse.lsp4j.Position;
+
+/**
+ * Prolog snippet context used to filter the xml processing instructions
+ * snippets.
+ * 
+ */
+public class PrologSnippetContext implements IXMLSnippetContext {
+
+	public static IXMLSnippetContext DEFAULT_CONTEXT = new PrologSnippetContext();
+
+	@Override
+	public boolean isMatch(ICompletionRequest request, Map<String, String> model) {
+		DOMDocument document = request.getXMLDocument();
+		// check if document already defined a xml processing instruction.
+		if (hasProlog(document)) {
+			return false;
+		}
+		Position start = request.getReplaceRange().getStart();
+		if (start.getLine() > 0) {
+			// The xml processing instruction must be declared in the first line.
+			return false;
+		}
+		// No xml processing instruction, check if completion was triggered before the
+		// document element
+		DOMElement documentElement = document.getDocumentElement();

```

Comment:
Can we also check for DocType? If I invoke completion at `<!DOCTYPE |root-element [`, I don't think I should get the prolog snippets.

![image](https://user-images.githubusercontent.com/20326645/81344693-458add00-9085-11ea-96b4-04b1ad052496.png)